File: algorithms/java/114_flatten_binary_tree_to_linked_list/Flatten.java
class Flatten {
    TreeNode prev = null;
    public void flatten(TreeNode root) {
        if (root == null) return;
        if (prev != null) {
            prev.right = root;
        }
        prev = root;
        TreeNode right = root.right;
        flatten(root.left);
        flatten(right);
        root.left = null;
    }
}

File: algorithms/cpp/114_flatten_bt_ll/Flatten.cpp
class Flatten {
public:
	/*
	* not using a globel pointer
	* link the curent right node to the current left node's right most node
	*/
	void flatten(TreeNode* root) {
        if (!root || (!root->left && !root->right))
            return;
        if (root->left) {
            flatten(root->left);
            
            TreeNode* tmp = root->right;
            root->right = root->left;
            root->left = NULL;
            
            TreeNode* cur = root->right;
            while (cur->right != NULL)
                cur = cur->right;
            cur->right = tmp;
        }
        flatten(root->right);
    }
};
